public/app/features/query/state/PanelQueryRunner.ts

```diff
@@ -98,8 +99,55 @@ export class PanelQueryRunner {
           let fieldConfig = this.dataConfigSource.getFieldOverrideOptions();
           let processFields = fieldConfig != null;
 
+          // TODO: create new "chunk" dataframe type ?
+          const firstFrame = data.series[0];
+          const isChunk = 'isChunk' in firstFrame && ((firstFrame as unknown) as { isChunk: boolean }).isChunk;
+
+          if (
+            isChunk &&
+            data.state === LoadingState.Streaming &&
+            processFields &&
+            processedCount > 0 &&
+            lastData.length &&
+            lastConfigRev === this.dataConfigSource.configRev
+          ) {
+            // Keep the previous field config settings
+            processedData = {
+              ...processedData,
+              state: LoadingState.Streaming,
+              // @ts-ignore
+              series: lastData.map((frame, frameIndex) => {
+                const newFields = frame.fields.map((field, fieldIndex) => {
+                  // @ts-ignore
+                  // TODO: should be fixed with a new dateFrame type
+                  data.series[frameIndex].fields[fieldIndex].values.forEach((val) => {
+                    (field.values as ArrayVector<unknown>).add(val);
+                  });
+                  return {
+                    ...field,
+                    state: {
+                      ...field.state,
+                      // TODO: recalculate
+                      calcs: undefined,
+                      range: undefined,
+                    },
+                  };
+                });
+
+                return {
+                  ...frame,
+                  length: lastData[frameIndex].length + data.series[frameIndex].length || undefined,
+                  fields: newFields,
+                };
+              }),
+            };
+            processFields = false;
+            sameStructure = true;
+          }
+
           // If the shape is the same, we can skip field overrides
           if (
+            // TODO: create CircularDataFrame on first `streaming` message
```

Comment:
Say we have 3 graphs subscribing to the same channel with different options, that is buffer size + filters:

graph1: 
- max buffer size: 100
- fields: [`a`, `b`]

graph2:
- max buffer size: 200
- fields: ['c', 'd']

graph3:
- max buffer size: 150
- fields: ['a', 'c', 'd']

The behavior on the main branch:
- 3 dataFrame buffers on the web worker in three different subscriptions. each buffer has different sizing and fields
- 3 dataFrame buffers on the main thread. -||-
- we transfer full-size dataFrame from the web worker to the main thread on each new data point


The proposed behavior in this PR:
- single buffer on the web worker with max size of any subscribed graph (in this case: 200) and all fields
- 3 full-size dataFrame buffers backed by a `CircularDataFrame` on the main thread. `CircularDataFrame` is there to ensure that we respect each graph's max length
- full-sized dataFrame is transferred only on the initial subscribe, otherwise we send only the appended values


does that make sense @grafana/grafana-edge-squad

Reply:
:thinking: CircularDataFrame may actually be a good candidate for the worker thread buffer, but is not great as a buffer in the query thread.

CircularDataFrame allocates a buffer, then just changes the index it writes to -- super fast/efficient... until something on the frontend (like the graph!) calls `toArray()` and it has to recreate the whole buffer!

StreamingDataFrame uses slice+push on a simple array and works very well with the graph

Reply:
@ArturWierzbicki -- i think your proposal sounds right, except you should use StramingDataFrame in *both* places.  in the main thread it will also check the max time in addition to max length!

Reply:
using `StreamingDataFrame` as discussed